rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+  - This also means that (multi-line) `with` expressions will use style #3 or #4, depending on their body.
+  ```nix
+  # Force-expand short attrset
+  meta = with lib; {
+    maintainers = [];
+  };
+  # Don't absorb since the body of `with pkgs;` is `with pyPkgs; ...`, which is not absorbable.
+  buildInputs =
+    with pkgs;
+    with pyPkgs;
+    [
+      some
+      dependencies
+    ];
+  ```
+
+**Alternatives**
+
+Function calls could always be absorbed. This would reduce indentation of their arguments in some cases. However, this may look really weird in other cases, especially when the binding is very long:
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+Consistent with this would be to also absorb `let` bindings and other expressions, however this might result in double indentation.
+
+```nix
+suff = let
+   foo = "bar"; # <-- double-indentation
+  in
+  foo;
+```
+
+#### Bindings semicolon placement
+
+The semicolon in bindings must always be placed on the same line as the expression it concludes.
+
+**Examples**
+
+```nix
+{
+  attr1 = bar;
+  attr2 = function call {
+    # stuff
+  };
+  attr3 =
+    function call
+      many
+      arguments;
+  attr4 =
+    let
+      foo = "bar";
+    in
+    some statement;
+  attr5 =
+    if foo then
+      "bar"
+    else
+      "baz";
+  attr6 =
+    let
+      foo = false;
+    in
+    if foo then "bar" else "baz";
+  attr7 = function (
+    if foo then
+      "bar"
+    else
+      "baz"
+  );
+  attr8 =
+    cond1
+    || cond2
+    ||
+      some function call
+      && cond3;
+}
+```
+
+**Alternatives**
+
+
+1. On a new line without indentation.
+  - This clearly marks a separation between attributes, however it is wasteful of space.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments
+  ;
+  attr3 =
+    let
+      foo = "bar";
+    in
+    some statements
+  ;
+  ```
+2. On a new line with one indentation level.
+  - Just as wasteful on space as (1), but a bit less clear about signaling the end of the binding.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments
+    ;
+  ```
+3. A mix of (1) and (2), where usually the semicolon is placed directly at the end of the binding.
+   But with exceptions in which the semicolon is placed onto the following line instead in cases where the value is a multiline `if` expression or nested operator.
+   These are the only syntax elements that may result in the semicolon being placed on a line with arbitrarily deep indentation.
+   
+   ```nix
+   attr4 =
+     if foo then
+       "bar"
+     else
+       "baz"
+   ;
+
+   attr5 =
+     let
+       foo = false;
+     in
+     foo || bar;
+       
+   attr7 =
+     cond1
+     || cond2
+     ||
+       some function call
+       && cond3
+   ;
+   ```
+
+### inherit
+
+The items must either be all on the same line, or all on a new line each (with indentation),
+in which case the semicolon must be on its own line with indentation.
```

Comment:
With bindings, the trade-off was between wasted vertical space and diff noise in some expressions (if, sometimes functions) where the preference fell to vertical space savings.

The same trade-off applies here but the preference went the other way which is inconsistent. I think that should explicitly be called out and have a reason given.

I assume this was done because appending elements to an inherit happens quite frequently (i.e. `inherit (lib) foo bar ...;`).

Reply:
Now added justification for this in https://github.com/NixOS/rfcs/pull/166/commits/f33b211ed60d0f12acad54bdfeecc0ca677a41fe :)